src/js/game/Entities/Agent.js

```diff
@@ -0,0 +1,156 @@
+const BaseEntity = require("./BaseEntity.js");
+const CallbackCommand = require("../CommandQueue/CallbackCommand.js");
+
+module.exports = class Player extends BaseEntity {
+  constructor(controller, type, x, y, name, isOnBlock, facing) {
+    super(controller, type, 'PlayerAgent', x, y, facing);
+    this.offset = (name === 'Agent') ? [-5, -15] : [-18, -32];
+    this.name = name;
+    this.isOnBlock = isOnBlock;
+    this.inventory = {};
+    this.movementState = -1;
+
+    if (controller.levelData.isEventLevel) {
+      this.moveDelayMin = 0;
+      this.moveDelayMax = 0;
+    } else {
+      this.moveDelayMin = 30;
+      this.moveDelayMax = 200;
+    }
+  }
+
+  // "Events" levels allow the player to move around with the arrow keys, and
+  // perform actions with the space bar.
+  updateMovement() {
+    if (!this.controller.attemptRunning || !this.controller.levelData.isEventLevel) {
+      return;
+    }
+    const queueIsEmpty = this.queue.isFinished() || !this.queue.isStarted();
+    const isMoving = this.movementState !== -1;
+    const queueHasOne = this.queue.currentCommand && this.queue.getLength() === 0;
+    const timeEllapsed = (+new Date() - this.lastMovement);
+    const movementAlmostFinished = timeEllapsed > 300;
+
+    if ((queueIsEmpty || (queueHasOne && movementAlmostFinished)) && isMoving) {
+      // Arrow key
+      if (this.movementState >= 0) {
+        let direction = this.movementState;
+        let callbackCommand = new CallbackCommand(this, () => { }, () => {
+          this.lastMovement = +new Date();
+          this.controller.moveDirection(callbackCommand, direction);
+        }, this.identifier);
+        this.addCommand(callbackCommand);
+        // Spacebar
+      } else {
+        let callbackCommand = new CallbackCommand(this, () => { }, () => {
+          this.lastMovement = +new Date();
+          this.controller.use(callbackCommand);
+        }, this.identifier);
+        this.addCommand(callbackCommand);
+      }
+    }
+  }
+
+  doMoveForward(commandQueueItem) {
+    var player = this,
+      groundType,
+      jumpOff,
+      levelModel = this.controller.levelModel,
+      levelView = this.controller.levelView;
+    let wasOnBlock = player.isOnBlock;
+    let prevPosition = this.position;
+    // update position
+    levelModel.moveForward();
+    // TODO: check for Lava, Creeper, water => play approp animation & call commandQueueItem.failed()
+
+    jumpOff = wasOnBlock && wasOnBlock !== player.isOnBlock;
+    if (player.isOnBlock || jumpOff) {
+      groundType = levelModel.actionPlane._data[levelModel.yToIndex(player.position[1]) + player.position[0]].blockType;
```

Comment:
Please use `levelModel.actionPlane.getBlockAt(position)` (one more below).

Reply:
Yup 👍 
again, that's just copy pasted from player.js
Once functionality is working I'm going to clean these files up.

Reply:
Sounds good!

Reply:
Also, just noticed.... this was copied over from Player.js but is this if/else really:
if(){
 groundType = levelModel.actionPlane.getBlockAt(player.position).blockType;
} else {
 groundType = levelModel.actionPlane.getBlockAt(player.position).blockType;
}

#ForWhatPurpose

Reply:
Oh, the else block is supposed to be referencing the _ground_ plane, not the _action_ plane